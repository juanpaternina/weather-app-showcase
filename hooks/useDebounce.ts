import { useRef } from 'react';

const useDebounce = (callback: Function, delay: number) => {
  const timeoutRef = useRef<NodeJS.Timeout | null>(null);

  const debouncedFunction = (...args: any[]) => {
    if (timeoutRef.current) {
      clearTimeout(timeoutRef.current);
    }
    timeoutRef.current = setTimeout(() => {
      callback(...args);
    }, delay);
  };

  return debouncedFunction;
};

export default useDebounce;
