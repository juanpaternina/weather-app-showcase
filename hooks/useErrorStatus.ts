import { useCallback, useState } from 'react';

export const useErrorStatus = () => {
  const [error, setError] = useState(false);
  const [errorMessage, setErrorMessage] = useState('');

  const updateErrorState = useCallback(
    (isError: boolean, lErrorMessage?: string) => {
      if (isError && errorMessage) {
        setErrorMessage(errorMessage);
        setError(isError);
        return;
      }

      if (isError && !lErrorMessage) {
        setErrorMessage('Something went wrong');
        return;
      }

      setError(isError);
      setErrorMessage('');
    },
    [],
  );

  return { error, errorMessage, updateErrorState };
};
