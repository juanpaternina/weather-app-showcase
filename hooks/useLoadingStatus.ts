import { useCallback, useState } from 'react';

export const useLoadingStatus = () => {
  const [loading, setLoading] = useState(false);

  const updateLoadingState = useCallback((isLoading: boolean) => {
    setLoading(isLoading);
  }, []);

  return { loading, updateLoadingState };
};
