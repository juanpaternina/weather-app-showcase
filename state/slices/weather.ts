import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { UserLocation } from '@/types/userLocation';

interface WeatherState {
  location: UserLocation;
}

const initialState: WeatherState = {
  location: {
    lat: 52.5167,
    lng: 13.4,
  },
};

const weatherSlice = createSlice({
  name: 'weather',
  initialState,
  reducers: {
    updateUserLocation: (state, action: PayloadAction<UserLocation>) => ({
      ...state,
      location: action.payload,
    }),
  },
});

export const { updateUserLocation } = weatherSlice.actions;
export default weatherSlice.reducer;
