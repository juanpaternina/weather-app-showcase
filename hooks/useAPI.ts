import { useState } from 'react';
import { useErrorStatus } from './useErrorStatus';
import { useLoadingStatus } from './useLoadingStatus';

export const useAPI = <T>() => {
  const { loading, updateLoadingState } = useLoadingStatus();
  const { error, errorMessage, updateErrorState } = useErrorStatus();

  const [data, setData] = useState<T | null>(null);

  const fetchAPI = (request: string) => {
    updateLoadingState(true);
    fetch(request)
      .then((response) => response.json())
      .then((data) => {
        setData(data);
      })
      .catch(() => {
        updateErrorState(true, "Couldn't fetch data, please try again later");
        setData(null);
      })
      .finally(() => {
        updateLoadingState(false);
      });
  };

  return {
    data,
    error,
    errorMessage,
    loading,
    fetchAPI,
  };
};
