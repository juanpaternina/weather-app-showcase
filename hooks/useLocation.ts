import { useCallback } from 'react';
import * as Location from 'expo-location';

import { useAppDispatch } from '@/state/store';
import { updateUserLocation } from '@/state/slices/weather';

export const useLocation = () => {
  const dispatch = useAppDispatch();

  const getLocation = useCallback(async () => {
    const { status } = await Location.requestForegroundPermissionsAsync();
    if (status !== 'granted') {
      return;
    }

    const { coords } = await Location.getCurrentPositionAsync({});

    dispatch(
      updateUserLocation({ lat: coords.latitude, lng: coords.longitude }),
    );
  }, [dispatch]);

  return { getLocation };
};
