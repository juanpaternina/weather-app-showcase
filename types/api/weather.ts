import { Day } from '../day';
import { Location } from '../location';
import { Weather } from '../weather';

export interface WeatherAPI {
  location: Location;
  current: Weather;
  forecast: Forecast;
}

export interface Forecast {
  forecastday: Forecastday[];
}

export interface Forecastday {
  date: string;
  date_epoch: number;
  day: Day;
}
