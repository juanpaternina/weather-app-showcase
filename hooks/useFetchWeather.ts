import { requestCreator } from '@/helpers/requestCreator';
import { API_TYPES } from '@/constants/Api';
import { useAPI } from './useAPI';
import { WeatherAPI } from '@/types/api/weather';

export const useFetchWeather = () => {
  const { data, error, errorMessage, loading, fetchAPI } = useAPI<WeatherAPI>();

  const fetchWeather = (lat: number, long: number) => {
    const requestURL = requestCreator(API_TYPES.FORECAST, {
      q: `${lat},${long}`,
      days: 5,
    });
    fetchAPI(requestURL);
  };

  return {
    data,
    error,
    errorMessage,
    loading,
    fetchWeather,
  };
};
