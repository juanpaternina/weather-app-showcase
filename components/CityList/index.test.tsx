import { render, fireEvent } from '@testing-library/react-native';
import { CityList } from './index';

import cities from '@/mocks/search.json';

import { useAppDispatch } from '@/state/hooks';
import { updateUserLocation } from '@/state/slices/weather';
import { useRouter } from 'expo-router';

jest.mock('@/state/hooks', () => ({
  useAppDispatch: jest.fn(),
}));

jest.mock('expo-router', () => ({
  useRouter: jest.fn(),
}));

jest.mock('@/state/slices/weather', () => ({
  updateUserLocation: jest.fn(),
}));

jest.mock('@expo/vector-icons/MaterialIcons', () => ({
  __esModule: true,
  default: ({ name, size, color }) => (
    <mock-icon
      data-testid={`material-icon-${name}`}
      size={size}
      color={color}
    />
  ),
}));

const mockDispatch = jest.fn();
const mockRouter = {
  back: jest.fn(),
};
const mockUpdateUserLocation = updateUserLocation;

describe('CityList Component', () => {
  beforeEach(() => {
    // const state = {
    //   weather: {
    //     userLocation: {
    //       lat: 51.509865,
    //       lng: -0.118092,
    //     },
    //   },
    // };

    jest.clearAllMocks();

    (useAppDispatch as unknown as jest.Mock).mockReturnValue(mockDispatch);
    (useRouter as unknown as jest.Mock).mockReturnValue(mockRouter);
    // (useAppSelector as unknown as jest.Mock).mockImplementation((f) =>
    //   f(state),
    // );
  });

  test('renders correctly with given text', () => {
    const { toJSON } = render(<CityList cities={cities} />);
    expect(toJSON()).toMatchSnapshot();
  });

  test('Should dispatch updateUserLocation when pressing a city', () => {
    const { getByTestId } = render(<CityList cities={cities.slice(0, 1)} />);
    const city = getByTestId('pressable-arrow-forward');
    fireEvent.press(city);

    expect(mockDispatch).toHaveBeenCalledWith(
      mockUpdateUserLocation({ lat: 51.509865, lng: -0.118092 }),
    );

    expect(mockRouter.back).toHaveBeenCalledTimes(1);
  });
});
