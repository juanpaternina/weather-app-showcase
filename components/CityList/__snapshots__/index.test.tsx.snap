// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CityList Component renders correctly with given text 1`] = `
<RCTScrollView
  data={
    [
      {
        "country": "Spain",
        "id": 714482,
        "lat": 40.4,
        "lon": -3.68,
        "name": "Madrid",
        "region": "Madrid",
        "url": "madrid-madrid-spain",
      },
      {
        "country": "Colombia",
        "id": 517469,
        "lat": 4.45,
        "lon": -74.16,
        "name": "Madrid",
        "region": "Cundinamarca",
        "url": "madrid-cundinamarca-colombia",
      },
      {
        "country": "Spain",
        "id": 714481,
        "lat": 39.47,
        "lon": -3.53,
        "name": "Madridejos",
        "region": "Castilla-La Mancha",
        "url": "madridejos-castilla-la-mancha-spain",
      },
      {
        "country": "Mexico",
        "id": 3325190,
        "lat": 19.08,
        "lon": -103.87,
        "name": "Madrid",
        "region": "Colima",
        "url": "madrid-colima-mexico",
      },
      {
        "country": "Philippines",
        "id": 1850841,
        "lat": 11.3,
        "lon": 123.73,
        "name": "Madridejos",
        "region": "Cebu",
        "url": "madridejos-cebu-philippines",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={[]}
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignContent": "center",
            "borderBottomWidth": 1,
            "borderColor": "black",
            "flexDirection": "row",
            "height": 50,
            "marginLeft": 15,
            "marginRight": 15,
            "paddingTop": 15,
            "verticalAlign": "center",
          }
        }
        testID="pressable-arrow-forward"
      >
        <Text
          style={
            {
              "alignItems": "center",
              "color": "black",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "fontSize": 16,
              "justifyContent": "center",
              "wordWrap": "break-word",
            }
          }
        >
          Madrid
          , 
          Madrid
          , 
          Spain
        </Text>
        <mock-icon
          color="black"
          data-testid="material-icon-arrow-forward"
          size={24}
        />
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignContent": "center",
            "borderBottomWidth": 1,
            "borderColor": "black",
            "flexDirection": "row",
            "height": 50,
            "marginLeft": 15,
            "marginRight": 15,
            "paddingTop": 15,
            "verticalAlign": "center",
          }
        }
        testID="pressable-arrow-forward"
      >
        <Text
          style={
            {
              "alignItems": "center",
              "color": "black",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "fontSize": 16,
              "justifyContent": "center",
              "wordWrap": "break-word",
            }
          }
        >
          Madrid
          , 
          Cundinamarca
          , 
          Colombia
        </Text>
        <mock-icon
          color="black"
          data-testid="material-icon-arrow-forward"
          size={24}
        />
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignContent": "center",
            "borderBottomWidth": 1,
            "borderColor": "black",
            "flexDirection": "row",
            "height": 50,
            "marginLeft": 15,
            "marginRight": 15,
            "paddingTop": 15,
            "verticalAlign": "center",
          }
        }
        testID="pressable-arrow-forward"
      >
        <Text
          style={
            {
              "alignItems": "center",
              "color": "black",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "fontSize": 16,
              "justifyContent": "center",
              "wordWrap": "break-word",
            }
          }
        >
          Madridejos
          , 
          Castilla-La Mancha
          , 
          Spain
        </Text>
        <mock-icon
          color="black"
          data-testid="material-icon-arrow-forward"
          size={24}
        />
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignContent": "center",
            "borderBottomWidth": 1,
            "borderColor": "black",
            "flexDirection": "row",
            "height": 50,
            "marginLeft": 15,
            "marginRight": 15,
            "paddingTop": 15,
            "verticalAlign": "center",
          }
        }
        testID="pressable-arrow-forward"
      >
        <Text
          style={
            {
              "alignItems": "center",
              "color": "black",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "fontSize": 16,
              "justifyContent": "center",
              "wordWrap": "break-word",
            }
          }
        >
          Madrid
          , 
          Colima
          , 
          Mexico
        </Text>
        <mock-icon
          color="black"
          data-testid="material-icon-arrow-forward"
          size={24}
        />
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignContent": "center",
            "borderBottomWidth": 1,
            "borderColor": "black",
            "flexDirection": "row",
            "height": 50,
            "marginLeft": 15,
            "marginRight": 15,
            "paddingTop": 15,
            "verticalAlign": "center",
          }
        }
        testID="pressable-arrow-forward"
      >
        <Text
          style={
            {
              "alignItems": "center",
              "color": "black",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "fontSize": 16,
              "justifyContent": "center",
              "wordWrap": "break-word",
            }
          }
        >
          Madridejos
          , 
          Cebu
          , 
          Philippines
        </Text>
        <mock-icon
          color="black"
          data-testid="material-icon-arrow-forward"
          size={24}
        />
      </View>
    </View>
  </View>
</RCTScrollView>
`;
